### ISSUE TITLE
Duplicate, or more, rom information labels in some cases, probably because of 'serials as id'

### ISSUE THREAD

**Issue by i30817** [libretro/RetroArch#7455]
# First and foremost consider this:
- Only RetroArch bugs should be filed here. Not core bugs or game bugs
- This is not a forum or a help section, this is strictly developer oriented

## Description

Some rom information that comes from libretro-database is being duplicated on RA rom information page.

for instance this is a megadrive playlist `Alien Soldier (Europe)`:

![screenshot from 2018-10-18 06-11-21](https://user-images.githubusercontent.com/757900/47132642-ef84f780-d29c-11e8-964e-33911f111c7d.png)

There are more hashsums listed but not visible (about 7 of them all different as far as i can tell). The top is also obviously a result of a merge of a different style of dat since it includes the serial.

The real rom has crc 0496e06c (the last listing on the screenshot and also the 'first' in a different style, of info, probably because take from another type of dat).

Other crcs (not listing md5 to simplify) are 0A9C04C9, 9B1A1D7F, B5CB1CEB, 3F5746BA, 3DB10C48, C4EB3F76

None of these other crcs appear on the original no-intro dat, but it's definitely the result of probably using serials to 'identify' games and ending up with multiple versions of the game on the 'info' instead of the *right* version.

grep on a repository of the libretro-database confirms it
```
grep -Eir "0A9C04C9|9B1A1D7F|B5CB1CEB|3F5746BA|3DB10C48|C4EB3F76"
metadat/no-intro/Sega - Mega Drive - Genesis.dat:	rom ( name "Alien Soldier (USA) (Virtual Console).md" size 2097152 crc 9B1A1D7F md5 54D06FE70C75BF6828A503BB8CE806B0 sha1 A7C6E7D6B3DF26A3CA381A170AD06EFAB0043C97 )
metadat/goodtools/GoodGen 3.21.dat:	rom ( name "Alien Soldier (E) [b1].bin" size 2097152 crc c4eb3f76 md5 72ada8ca42e600ed2128b054af4fc88e sha1 426cbeb947e5a9fe9aac2d3288572cf439d45c4a )
metadat/goodtools/GoodGen 3.21.dat:	rom ( name "Alien Soldier (E) [b2].bin" size 2097664 crc 3db10c48 md5 9847bf660e0effe7f3e6aa971447ed28 sha1 d95a87992705e041d89abcbe9c26194070fcc4c9 )
metadat/goodtools/GoodGen 3.21.dat:	rom ( name "Alien Soldier (E) [f1].bin" size 2097152 crc 3f5746ba md5 328592e16e1a380f6123e0b444956c85 sha1 316dbce9d6d17606d01498b5bd7a5e28dba72f7e )
metadat/goodtools/GoodGen 3.21.dat:	rom ( name "Alien Soldier (E) [f2].bin" size 2097152 crc b5cb1ceb md5 1d577d96dbf03b13babdfd91e163e9a8 sha1 cd29bb5b9a727148d9e3601507187db2dd02cd2d )
metadat/goodtools/GoodGen 3.21.dat:	rom ( name "Alien Soldier (E) [f3].bin" size 2097152 crc 9b1a1d7f md5 54d06fe70c75bf6828a503bb8ce806b0 sha1 a7c6e7d6b3df26a3ca381a170ad06efab0043c97 )
metadat/goodtools/Sega - Mega Drive - Genesis.dat:	rom ( name "Alien Soldier (E) [b1].bin" size 2097152 crc c4eb3f76 md5 72ada8ca42e600ed2128b054af4fc88e sha1 426cbeb947e5a9fe9aac2d3288572cf439d45c4a )
metadat/goodtools/Sega - Mega Drive - Genesis.dat:	rom ( name "Alien Soldier (E) [b2].bin" size 2097664 crc 3db10c48 md5 9847bf660e0effe7f3e6aa971447ed28 sha1 d95a87992705e041d89abcbe9c26194070fcc4c9 )
metadat/goodtools/Sega - Mega Drive - Genesis.dat:	rom ( name "Alien Soldier (E) [f1].bin" size 2097152 crc 3f5746ba md5 328592e16e1a380f6123e0b444956c85 sha1 316dbce9d6d17606d01498b5bd7a5e28dba72f7e )
metadat/goodtools/Sega - Mega Drive - Genesis.dat:	rom ( name "Alien Soldier (E) [f2].bin" size 2097152 crc b5cb1ceb md5 1d577d96dbf03b13babdfd91e163e9a8 sha1 cd29bb5b9a727148d9e3601507187db2dd02cd2d )
metadat/goodtools/Sega - Mega Drive - Genesis.dat:	rom ( name "Alien Soldier (E) [f3].bin" size 2097152 crc 9b1a1d7f md5 54d06fe70c75bf6828a503bb8ce806b0 sha1 a7c6e7d6b3df26a3ca381a170ad06efab0043c97 )

```

Anyway, it's extremely sketchy to use serials to 'identify' roms *especially* if RA is going to try to merge all possible dumps on the database - not that only using no-intro would help, for instance the first on that list is the no-intro virtual console version).

### Expected behavior
Stop using serials to uniquely identify games. Use them as shared data foreign keys (like pictures of covers) but not for 'actually unique' information like hashcodes or other properties on the per-game entries on dats. A single game should give a single result, end of story.

### Actual behavior

Trying to accommodate every rom version and *not* calculate hashcodes one time 'because it's slower to scan' leads to this.

### Steps to reproduce the bug

1. [First step]
2. [Second step]
3. [and so on...]

### Bisect Results

[Try to bisect and tell us when this started happening]

### Version/Commit
You can find this information under Information/System Information

- RetroArch: [version/commit]

### Environment information

- OS: [The operating system you're running]
- Compiler: [In case you are running local builds]

**Comment by twinaphex** [comment]
This is a fair point. These duplicate entries need a solution of some sorts.

**Comment by i30817** [comment]
There is also the case where two or more different dump groups have the same exact rom (hashcode too obviously). In that case my recommendation is to ... let the more popular/respected group win the race (ie: no intro would always win over goodtools).

In fact i'd suggest duplicated entries like that don't even get into the database, just one canonical for the most respected group atm.

**Comment by i30817** [comment]
You know what, i'm going to give a suggestion on the assumption you're not going to decide to revert the serial scanning.

Give a option, either:

1.  have serial scans and disable features that need hashcodes or calculate them lazily (before caching  the playlist obviously). This means from what i can tell conservatively: netplay, the proposed feature for opening the roms webpages of a hack, disable the 'information' page entries that refer to crcs and keep only things like 'publisher', 'release date', 'tags' etc.

You or a 3rd party would still be able to bundle game specific configuration bundles (say for fixing letterboxing in ps1 games or having arcade cabinet overlays) but they'd would have false positives for certain hacks (which would require 'specific' handling overriding, ie: if the crc/md5 is known, overlay these changes).

Basically don't assume that just because you have a serial match on the database that the hashcoded and the same and think case by case if the operation requires a unique rom.

2. Or the user changes a setting to make rom scan use hashcodes (as they should) and every feature is enabled from the start at the cost of much slower scan for large collections.

Leaving things 'lazy' on 1 might be a acceptable result, if you put in a little warning on the user trying a feature requiring exact matches on roms that are 'large enough' on size 'this feature requires a long crc calculation Y/N?'

**Comment by i30817** [comment]
Notice also that the serial is not a perfect for a fuzzy key either, often different regions or re-editions of a game have very different serials. Except the wii/gamecube that has a suffix for region iirc (dolphin takes advantage of this).

**Comment by i30817** [comment]
The reddit discussion also made me realize something else. Having a serial scanner is even worse for 'security' than accepting hacks with the right crcs (not as bad as listing anything with the right extension though).

Any ROM that has the right serial could be confused with a whitelisted version of the game, when if you kept using the hashes, they'd at least have to put a malicious rom on romhacking.net.

**Comment by retro-wertz** [comment]
have these of fbalpha, nes, snes and others. considering playlists are created based on hash, the other hashes should not show. (posting here to config this issue)

![screenshot_2018-12-22_22-36-01](https://user-images.githubusercontent.com/23315382/50375573-52e32b00-063a-11e9-8e7f-18dbe47c497b.png)

**Comment by orbea** [comment]
I noticed this too, it seems dependent on the game more than the core as some games won't cause it.

**Comment by ner00** [comment]
This is pretty backward and weird that it hasn't been addressed for so long. Just don't use the serial as a UID, it makes no sense because it doesn't work.

**Comment by i30817** [comment]
It's fast and that's all that matters to users. Until there is a option to opt-out of the insanity this will still happen.

**Comment by ner00** [comment]
> It's fast and that's all that matters to (l)users.

It depends, doesn't it? Even some users scanning folders full with gigabytes of isos may be more inclined to be patient and have their collection properly arranged, but I understand the performance benefit - especially if you're going to scan a folder multiple times. 

> Until there is a option to opt-out of the insanity this will still happen.

An opt-out option at the very least would be nice. At least for single file scanning it could consider calculating a checksum by default (with serial as opt-in), and for folder-wide scan being opt-out.

I'm curious to see some benchmarks on crc32 across platforms and file sizes, it will obviously be a whole lot slower than seeking a few bytes at an offset for sure.

**Comment by i30817** [comment]
oh, it takes hours, depending if you have a 10 years+ harddisk instead of a ssd. But with the magic of fileystem extended attributes you *could* only do it once. And as a bonus you could even make softpatches record the 'correct' hardpatched hash, with some trickery.

It doesn't work for hardlinks unfortunately, because linux was stupid here.
